File: pathplanner/image/animate.py
from typing import Mapping, Sequence

from PIL import Image, ImageDraw

from image.draw import draw_path


def in_bounds(r: int, c: int, height: int, width: int) -> bool:
    return 0 <= r < height and 0 <= c < width


TraceSnapshot = Mapping[str, object]


def render_search_gif(
    base_img: Image.Image,
    trace: Sequence[TraceSnapshot],
    final_path: Sequence[tuple[int, int]] | None,
    output_path: str,
    *,
    max_frames: int = 200,
    frame_duration_ms: int = 50,
    final_frame_duration_ms: int = 700,
) -> None:
    if not trace:
        return

    canvas = base_img.convert("RGB") if base_img.mode not in {
        "RGB", "RGBA"} else base_img.copy()
    width, height = canvas.size

    stride = max(1, len(trace) // max_frames) if len(trace) > max_frames else 1
    frames: list[Image.Image] = []
    durations: list[int] = []

    blocked_seen: set[tuple[int, int]] = set()

    for index in range(0, len(trace), stride):
        snapshot = trace[index]
        frame = canvas.copy()
        draw = ImageDraw.Draw(frame)

        for (r, c) in snapshot.get("blocked", ()):
            if not in_bounds(r, c, height, width):
                continue
            blocked_seen.add((r, c))
            draw.point((c, r), fill="black")

        for (r, c) in snapshot.get("visited", ()):
            if not in_bounds(r, c, height, width):
                continue
            draw.point((c, r), fill="blue")

        for (r, c) in snapshot.get("open", ()):
            if not in_bounds(r, c, height, width):
                continue
            draw.point((c, r), fill="yellow")

        current = snapshot.get("current")
        if current:
            r, c = current
            if in_bounds(r, c, height, width):
                draw.point((c, r), fill="lime")

        frames.append(frame)
        durations.append(frame_duration_ms)

    final_blocked = []
    for cell in blocked_seen:
        if in_bounds(cell[0], cell[1], height, width):
            final_blocked.append(cell)

    final_frame = draw_path(canvas, list(final_path or []), final_blocked)
    frames.append(final_frame)
    durations.append(final_frame_duration_ms)

    frames[0].save(
        output_path,
        save_all=True,
        append_images=frames[1:],
        format="GIF",
        duration=durations,
        loop=0,
    )

File: rustfinder/src/image/animate.rs
use std::collections::HashSet;
use std::fs::File;
use std::io::BufWriter;

use anyhow::Result;
use image::codecs::gif::{GifEncoder, Repeat};
use image::{Delay, DynamicImage, Frame, Rgb, RgbImage, RgbaImage};

use crate::finder::messaging::SearchStep;

fn make_frame(image: RgbaImage, delay_ms: u16) -> Frame {
    let delay = Delay::from_numer_denom_ms(delay_ms as u32, 1);
    Frame::from_parts(image, 0, 0, delay)
}

fn in_bounds(row: usize, col: usize, height: usize, width: usize) -> bool {
    row < height && col < width
}

pub fn render_search_gif(
    base_img: &DynamicImage,
    trace: &[SearchStep],
    final_path: Option<&[(usize, usize)]>,
    output_path: &str,
    max_frames: usize,
    frame_duration_ms: u16,
    final_frame_duration_ms: u16,
) -> Result<()> {
    if trace.is_empty() {
        return Ok(());
    }

    let canvas: RgbImage = base_img.to_rgb8();
    let (width, height) = canvas.dimensions();
    let width_usize = width as usize;
    let height_usize = height as usize;

    let stride = if max_frames > 0 && trace.len() > max_frames {
        let step = trace.len() / max_frames;
        if step == 0 { 1 } else { step }
    } else {
        1
    };

    let mut frames: Vec<Frame> = Vec::new();
    let mut blocked_seen: HashSet<(usize, usize)> = HashSet::new();

    for index in (0..trace.len()).step_by(stride) {
        let snapshot = &trace[index];
        let mut frame_img = canvas.clone();

        for &(row, col) in &snapshot.blocked {
            if row < 0 || col < 0 {
                continue;
            }
            let (r, c) = (row as usize, col as usize);
            if !in_bounds(r, c, height_usize, width_usize) {
                continue;
            }
            blocked_seen.insert((r, c));
            frame_img.put_pixel(c as u32, r as u32, Rgb([0, 0, 0]));
        }

        for &(row, col) in &snapshot.visited {
            if !in_bounds(row, col, height_usize, width_usize) {
                continue;
            }
            frame_img.put_pixel(col as u32, row as u32, Rgb([0, 0, 255]));
        }

        for &(row, col) in &snapshot.open_nodes {
            if !in_bounds(row, col, height_usize, width_usize) {
                continue;
            }
            frame_img.put_pixel(col as u32, row as u32, Rgb([255, 255, 0]));
        }

        let (row, col) = snapshot.current;
        if in_bounds(row, col, height_usize, width_usize) {
            frame_img.put_pixel(col as u32, row as u32, Rgb([50, 205, 50]));
        }

        frames.push(make_frame(
            DynamicImage::ImageRgb8(frame_img).into_rgba8(),
            frame_duration_ms,
        ));
    }

    let final_blocked: Vec<(isize, isize)> = blocked_seen
        .into_iter()
        .map(|(r, c)| (r as isize, c as isize))
        .collect();

    let base_dynamic = DynamicImage::ImageRgb8(canvas.clone());
    let final_image = crate::image::draw::draw_path(
        &base_dynamic,
        final_path.unwrap_or(&[]),
        &final_blocked,
    );
    frames.push(make_frame(
        final_image.into_rgba8(),
        final_frame_duration_ms,
    ));

    let file = File::create(output_path)?;
    let writer = BufWriter::new(file);
    let mut encoder = GifEncoder::new(writer);
    encoder.set_repeat(Repeat::Infinite)?;
    for frame in frames {
        encoder.encode_frame(frame)?;
    }

    Ok(())
}
